src/python/pants/backend/native/targets/native_artifact.py

```diff
@@ -1,14 +1,34 @@
 # Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+from dataclasses import FrozenInstanceError, dataclass
 from hashlib import sha1
+from typing import Any
 
 from pants.base.payload_field import PayloadField
-from pants.util.objects import datatype
 
 
-class NativeArtifact(datatype(['lib_name']), PayloadField):
+# NB: We manually implement __hash__(), rather than using unsafe_hash=True, because PayloadField
+# will mutate _fingerprint_memo and we must ensure that mutation does not affect the hash. For
+# extra safety, we override __setattr__() to ensure that the hash can never be accidentally changed.
```

Comment:
Does this apply to all `PayloadField`s? Does every other `PayloadField` also have to check if it's frozen? If so, could we at least at TODO here to turn this into a mixin or decorator to avoid error-prone boilerplate?

Reply:
Good question. This is the only `PayloadField` that used `datatype()`. The others do a mix of extending `tuple`, `OrderedSet`, or implementing their own `__hash__()`. 

I'm going to hold off on the TODO because this is a bigger question for what to do with `PayloadField`. For example, should all of the implementing classes become `@dataclass`es? Could we rewrite `PayloadField` to be immutable so that we can stop worrying about this `__setattr__()` concern? I think a GitHub issue could be appropriate to open!